### DISCUSSION TITLE
Getting Image to CTK label from URL

### DISCUSSION THREAD

**Discussion by Kirknasty54** [TomSchimansky/CustomTkinter#2169] [Q&A] [answered]
So I'm working on a library management system that was based on a final for one of my classes, and right now I'm trying to implement a search functionality for members so they can search up by book title, author, or isbn13 number. What I'm wanting to happen is that when users click the search button, I execute a sqlite3 query to pull all books from my database that some aspect that matches what the user put in the search bar. When the books are pulled, the first book's image should be pulled up. This is what I'm having trouble with. Actually getting the image to load onto the label. I'm using google books api as that seemed to be the simplest solution. 
Here is all my related code (also this project is on my repos if you want a fuller picture look)
```
`        def onSearchBtnClicked(event):
            print('this function was called')
            if search_entry:
                search_text = search_entry.get()
                with sqlite3.connect('book_list.db') as conn:
                    cursor = conn.cursor()
                    sql_query = f"SELECT * FROM book_registery WHERE title LIKE '%{search_text}%' OR authors LIKE '%{search_text}%' OR isbn13 LIKE '%{search_text}%'"
                    cursor.execute(sql_query)
                    rows = cursor.fetchall()
                    for row in rows:
                        cover_img = Book.loadCover(row[6])
                        if cover_img:
                            book_cover = ctk.CTkImage(dark_image=Image.open(cover_img), size=(48, 48))
                            book_img.config(image=cover_img)
                            book_img.image = cover_img
                            print(row[6])
                        else: print(f'cover img not loaded for {row[6]}')`
```
````
@staticmethod
    def getCover(isbn):
        base_url = 'https://www.googleapis.com/books/v1/volumes'
        params = {'q': f'isbn:{isbn}'}
        response = requests.get(base_url, params=params)
        data = response.json()
        if 'items' in data and data['items']:
            volume_info = data['items'][0]['volumeInfo']
            cover_url = volume_info['imageLinks']['thumbnail'] if 'imageLinks' in volume_info else None
            return cover_url
        else:
            print(f"Error: {response.status_code}")
        return None
```

    @staticmethod
    def loadCover(isbn):
        cover_url = Book.getCover(isbn)
        if cover_url:
            response = requests.get(cover_url)
            if response.status_code == 200:
                img = response.content
                img = Image.open(BytesIO(img))
                coverImg = ImageTk.PhotoImage(img)
                return coverImg
            else:
                print(f'error getting img status code {response.status_code}')
        else: print('cover not found ')`

**Comment by dipeshSam** [comment]
> So I'm working on a library management system that was based on a final for one of my classes, and right now I'm trying to implement a search functionality for members so they can search up by book title, author, or isbn13 number. What I'm wanting to happen is that when users click the search button, I execute a sqlite3 query to pull all books from my database that some aspect that matches what the user put in the search bar. When the books are pulled, the first book's image should be pulled up. This is what I'm having trouble with. Actually getting the image to load onto the label. I'm using google books api as that seemed to be the simplest solution. 
> Here is all my related code (also this project is on my repos if you want a fuller picture look)
> ```
> `        def onSearchBtnClicked(event):
>             print('this function was called')
>             if search_entry:
>                 search_text = search_entry.get()
>                 with sqlite3.connect('book_list.db') as conn:
>                     cursor = conn.cursor()
>                     sql_query = f"SELECT * FROM book_registery WHERE title LIKE '%{search_text}%' OR authors LIKE '%{search_text}%' OR isbn13 LIKE '%{search_text}%'"
>                     cursor.execute(sql_query)
>                     rows = cursor.fetchall()
>                     for row in rows:
>                         cover_img = Book.loadCover(row[6])
>                         if cover_img:
>                             book_cover = ctk.CTkImage(dark_image=Image.open(cover_img), size=(48, 48))
>                             book_img.config(image=cover_img)
>                             book_img.image = cover_img
>                             print(row[6])
>                         else: print(f'cover img not loaded for {row[6]}')`
> ```
> ````
> @staticmethod
>     def getCover(isbn):
>         base_url = 'https://www.googleapis.com/books/v1/volumes'
>         params = {'q': f'isbn:{isbn}'}
>         response = requests.get(base_url, params=params)
>         data = response.json()
>         if 'items' in data and data['items']:
>             volume_info = data['items'][0]['volumeInfo']
>             cover_url = volume_info['imageLinks']['thumbnail'] if 'imageLinks' in volume_info else None
>             return cover_url
>         else:
>             print(f"Error: {response.status_code}")
>         return None
> ```
> 
>     @staticmethod
>     def loadCover(isbn):
>         cover_url = Book.getCover(isbn)
>         if cover_url:
>             response = requests.get(cover_url)
>             if response.status_code == 200:
>                 img = response.content
>                 img = Image.open(BytesIO(img))
>                 coverImg = ImageTk.PhotoImage(img)
>                 return coverImg
>             else:
>                 print(f'error getting img status code {response.status_code}')
>         else: print('cover not found ')`

@Kirknasty54 Could you please elaborate a bit what problem exactly you are facing while showing an image as label? Error message, screenshot may help.

**Reply by Kirknasty54** [reply]
Sure thing, sorry. I should have thought of that 
![ctkprojecterror](https://github.com/TomSchimansky/CustomTkinter/assets/93236436/4e68d404-0ab6-4779-aef0-d87f4a6f623d)

**Comment by dipeshSam** [comment, marked as answer]
@Kirknasty54, The problem is most likely due to the image handling in Byte streams and converting the result to `CTkImage` else. Make sure that `cover_url = volume_info['imageLinks']['thumbnail'] if 'imageLinks' in volume_info else None` gives correct and reachable URL of the image, you can test with `params` if URL will not be reachable or check your API implementations.

**A class is made to fulfill the purpose easily.** Use this class to handle web images in `customtkinter` or consider its implementation in your cases. The class is simple, fast, and efficient.

**Video demonstration of the class usage:**
https://github.com/TomSchimansky/CustomTkinter/assets/72375763/2702e261-5b02-492b-b921-eab335394c74

**Code:**
``` Python
from customtkinter import (
    CTk,
    CTkLabel,
    CTkFont,
    CTkImage
)

from requests import get as get_from
from PIL.Image import open as open_img
from io import BytesIO

class CTkUrlLabel(CTkLabel):
    """
    A class designed to ease and facilitate the image fetching from web URL in `customtkinter`.

    Specific methods:
    - cget()        : Updated to get the url image.
    - configure()   : Updated to update the url image.

    Specific attributes:
    - url           : Complete url of the image
    - params        : Some webservers requires params while some do not.
    - url_image_size: If None, it will use actual size of the image

    """

    def __init__(self,
        master: any,
        width: int = 0,
        height: int = 28,
        corner_radius: int | None = None,
        bg_color: str | tuple[str, str] = "transparent",
        fg_color: str | tuple[str, str] | None = None,
        text_color: str | tuple[str, str] | None = None,
        text_color_disabled: str | tuple[str, str] | None = None, text: str = "CTkLabel",
        font: tuple | CTkFont | None = None,
        image: CTkImage | None = None,
        compound: str = "center",
        anchor: str = "center",
        wraplength: int = 0,
        url: str | None = None,
        params: dict | None = None,
        url_image_size: tuple[int, int] = None,
        **kwargs):

        # Saving URL Image specific options
        self._url            = url
        self._url_image      = None
        self.__image         = None
        self._params         = params
        self._url_image_size = url_image_size

        if url:
            self.__image, image = image, None

        super().__init__(master, width, height, corner_radius, bg_color, fg_color, text_color, text_color_disabled, text, font, image, compound, anchor, wraplength, **kwargs)
            
        # Calling the URL function at start
        if url: self._fetch_from_url()

    def _fetch_from_url(self):
        try:
            # Fetching the image from resource locator
            url_response = get_from(url=self._url, params=self._params)
            url_image = open_img(BytesIO(url_response.content))

            # Setting up the new fetched image.
            self._url_image = CTkImage(url_image, None,
                size=url_image.size if not self._url_image_size else self._url_image_size)
            self.configure(image=self._url_image)

        except:
            # If there is an error, set default image.
            self.configure(image=self.__image)

    def cget(self, attribute_name: str) -> any:
        """On 'url_image' returns object of `CTkImage` if any."""
        if "url_image" == attribute_name:
            return self._url_image

        return super().cget(attribute_name)
    
    def configure(self, require_redraw=False, **kwargs):
        """URL image options: 'url', 'params', and 'url_image_size'."""
        if "url" in kwargs:
            self._url = kwargs.pop("url")
            self._fetch_from_url()
        if "params" in kwargs:
            self._params = kwargs.pop("params")
            self._fetch_from_url()
        if "url_image_size" in kwargs:
            self._url_image_size = kwargs.pop("url_image_size")
            self._fetch_from_url()

        return super().configure(require_redraw, **kwargs)


# Complex sample usage:
from customtkinter import (
    CTkEntry, CTkButton, CTkRadioButton, StringVar)
from threading import Thread

if __name__ == "__main__":
    urls: dict = {
        "url1" : "https://camo.githubusercontent.com/98e693d243cde46579166501d14a9259daa92de146d9ba7d6af25255c4707974/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f696d616765732f6d6f64756c65732f70726f66696c652f616368696576656d656e74732f7374617273747275636b2d64656661756c742e706e67",
        "url2" : "https://camo.githubusercontent.com/d4a52cfb3181d13b7a8267394f66299c39d3c9c1da254f6beea092af1c755cca/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f696d616765732f6d6f64756c65732f70726f66696c652f616368696576656d656e74732f70756c6c2d736861726b2d64656661756c742e706e67"
    }

    def on_fetch():
        def on_target():
            try:
                button.configure(text="Fetching...", state="disabled")
                image_label.configure(url=url, text=name, compound="left")
                button.configure(text="Fetch", state="normal")
            except:
                button.configure(text="Fetch", state="normal")
                image_label.configure(text="Error: Check the network connection or the url specified.")

        if not entry.get():
            name = "Startruck" if avail_urls.get() == urls["url1"] else "Pull Shark"
            url  = avail_urls.get()
        else:
            name = image_name.get()
            url = entry.get()

        # Going through thread to prevent GUI freezing
        Thread(target=on_target).start()

        
    app = CTk()
    app.wm_title("CTkUrlLabel Test")
    app.geometry("700x500")
    app.grid_columnconfigure((0, 1), weight=1)

    # Giving header message
    CTkLabel(app, text="Enter the image url or use available one.", font=CTkFont(weight="bold", size=15)).grid(columnspan=2, pady=10)

    # Taking image URL
    entry = CTkEntry(app, placeholder_text="Enter full image url")
    entry.grid(row=1, column=0, sticky="ew", padx=10, pady=10)

    image_name = CTkEntry(app, placeholder_text="Enter display name of the image")
    image_name.grid(row=1, column=1, sticky="ew", padx=10, pady=10)

    # Placing available options
    avail_urls = StringVar(value=urls["url1"])
    CTkRadioButton(app, value=urls["url1"], variable=avail_urls, text="Github Startruck").grid(row=2, column=0)
    CTkRadioButton(app, value=urls["url2"], variable=avail_urls, text="Github Pull Shark").grid(row=2, column=1)

    # Giving a fetch image button
    button = CTkButton(app, text="Fetch", command=on_fetch)
    button.grid(row=3, column=0, columnspan=2, pady=10)
    CTkLabel(app, text="Fetched image:", font=CTkFont(weight="bold", size=15)).grid(row=4, column=0)

    # The CTkUrlImage demonstration
    image_label = CTkUrlLabel(app, text="[Image will be loaded here]")
    image_label.grid(row=5, column=0, columnspan=2, pady=10)
    
    app.mainloop()
```
**In the code:**
- `CTkUrlLabel` works exactly such as `CTkLabel` does.
- `url` parameter added.
- `params` parameter added (useful in some cases).
- `url_image_size` added for custom sizing of new image.
- If URL is not reachable, the default `image` can be used as place holder image.
- `cget()` and `configure()` methods are updated to update the `CTkUrlLabel` any time.
- It is advised to use Thread while fetching an image to prevent GUI freezing.

***If you want to use it simply, you can.***
Let you have saved the class `CTkUrlLabel` into a file named as `ctk_url_label.py` in your current working directory. Then a simple/basic use case example:
``` Python
from customtkinter import CTk
from ctk_url_label import CTkUrlLabel

if __name__ == "__main__":
    app = CTk()
    app.geometry("500x350")

    image_url = "https://cdn.kobo.com/book-images/f5f51626-0ab5-4d16-9ece-f8e256b5dc05/1200/1200/False/the-illustrated-theory-of-everything.jpg"

    image_label = CTkUrlLabel(app, text="  Theory Of Everything\n  by Stephen Hawking", compound="left", url=image_url, url_image_size=(35, 50))
    image_label.grid(padx=100, pady=100)
    
    app.mainloop()
```
***Output:*** 
![simple_use_case_example_of_ctk_url_label](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/3b7f45e7-5a79-47c0-be0a-6220b30d4a89)


**Important:**
Please do let me know whether it works well or not. If it produces errors, or you encounter any bug, inform them here.

**Hope, it will be helpful for you.**
*Happy customtkinter :)*

**Reply by Kirknasty54** [reply]
I've been toying with it, and i think your suggestion was the exact thing I needed. You are a life saver, this has been driving me nuts. Thank you so much! One last question. How do you create custom shaped buttons? Because this is my first cutomtkiner or tkinter project and I've been having some trouble figuring out how to create custom shaped/image buttons. So what I've been doing is just creating a label and binding it to a left click so that I can more easily give it a custom image

**Comment by dipeshSam** [comment]
> I've been toying with it, and i think your suggestion was the exact thing I needed. You are a life saver, this has been driving me nuts. Thank you so much! One last question. How do you create custom shaped buttons? Because this is my first cutomtkiner or tkinter project and I've been having some trouble figuring out how to create custom shaped/image buttons. So what I've been doing is just creating a label and binding it to a left click so that I can more easily give it a custom image

@Kirknasty54, **Glad to hear that it is helpful for you. You are most welcome.**
To create an image button or icon button, or you've described it as custom shaped button, there are standard three ways to achieve them in `Tkinter` or `customtkinter`. Here is description for **customtkinter**:

1. Provide `image` parameter to the `CTkButton`. It natively supports image as `CTkImage` object.
2. Provide `image` parameter to the `CTkLabel`. Bind click event.
3. Using `CTkCanvas`. Draw shapes and bind click event.

**Video demonstration:** (The video has two purposes: first, shows cursor change effect. Second, shows **Label change** on click.)
![different_btns](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/bb7ac388-a7ed-490f-b878-0bf0138b8d96)

**Code:**
``` Python
from customtkinter import (
    CTk,
    CTkButton,
    CTkLabel,
    CTkImage,
    CTkCanvas,
    CTkFont
)
from customtkinter import set_appearance_mode
from PIL.Image import open as open_img

def getStarVerteces(scaling_factor: float = 1):
    star_verteces = [(0, 30), (30, 30), (40, 0), (50, 30), (80, 30),
                      (55, 50), (65, 80), (40, 60), (15, 80), (25, 50)]
    return [(vertex[0]*scaling_factor, vertex[1]*scaling_factor) for vertex in star_verteces]

if __name__ == "__main__":
    app = CTk()
    app.geometry("500x350")
    app.grid_columnconfigure(0, weight=1)
    # set_appearance_mode("light")

    label = CTkLabel(app, text="This is button demonstration", font=CTkFont(weight="bold", size=20))
    label.grid(pady=20)

    # Image button using Button itself.    
    icon = CTkButton(app,   text="", command=lambda: label.configure(text="Search button clicked!"),
            image=CTkImage(light_image=open_img("icon_search_light.png"), dark_image=open_img("icon_search_dark.png"), size=(25, 25)),
            fg_color="transparent", hover=False)
    icon.grid(row=1, pady=20)


    # Image button using Label
    icon = CTkLabel(app,    text="", cursor="hand2",
            image=CTkImage(light_image=open_img("icon_history_light.png"), dark_image=open_img("icon_history_dark.png"), size=(25, 25)))
    icon.grid(row=2, pady=20)

    icon.bind("<Button-1>", lambda _: label.configure(text="History button clicked!"))

    # Image button using Canvas. (Different shaped button)
    # Setting up canvas
    canvas = CTkCanvas(app, width=50, height=50, highlightthickness = 0, cursor="hand2",
            bg=app._fg_color[ 0 if app._get_appearance_mode() == "light" else 1])
    canvas.grid(row=3, pady=20)

     # Drawing an star shape
    canvas.create_polygon(getStarVerteces(scaling_factor=0.6), fill=canvas["bg"],
            outline="black" if app._get_appearance_mode() == "light" else "white", tags="star_button")
            
     # Making the star shape clickable
    canvas.tag_bind("star_button", "<Button-1>", lambda _: label.configure(text="Star button clicked!"))
    
    app.mainloop()
```
**In the coed:**
- You can set foreground/background colors for both `CTkButton` and `CTkLabel`.
- Text can also be added and can be positioned with respect to image using `compound` parameter.
- I drawn Star (facilitated scaling), and set no color. If you want you can set border color, fill color etc.
- For hover effect, you can bind `"<Enter>"` and `"<Leave>"` events and can configure colors accordingly.
- In case of `CTkButton`, to set width and height according to the image, set them as 0 initially.
- In case of `CTkButton`, to set the corner radius, set width and height greater than the image size.
- The provided code works in both the modes light and dark one.
- I also set, cursor from "arrow" (default) to as "hand2" (shows clickable situation). You can modify it.

**Output (in different appearance modes):**
![different_btns_1](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/d78f16a8-d27d-438d-923a-45828e19ba6b) ![different_btns](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/93919228-2e77-4733-9809-cd076a75de45)

**Assets:**
The icons I've used in the code are:
1. **Search icon**: 
        Light: ![Icon_Search_Light](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/48ca7a85-6421-406f-ba41-a62e3bc25db3)
        Dark: ![Icon_Search_Dark](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/56340f68-002c-4ae0-91a8-5171a1981406)
2. **History icon**:
        Light: ![Icon_History_Light](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/04e102ff-d7c1-4ab2-8e13-c949db770a0d)
        Dark: ![Icon_History_Dark](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/611d0b21-bd79-4ddf-a38c-a17d538441e1)



**Hope, it will be helpful for you.**
***Happy customtkinter :)***

**Reply by Kirknasty54** [reply]
You are a godsend my friend, thank you so much for your help!

**Reply by dipeshSam** [reply]
> You are a godsend my friend, thank you so much for your help! 

@Kirknasty54 **You are most welcome brother :)**
